src/lib/rng/auto_rng/auto_rng.cpp

```diff
@@ -83,27 +87,24 @@ std::string AutoSeeded_RNG::name() const
    return m_rng->name();
    }
 
-void AutoSeeded_RNG::add_entropy(const uint8_t in[], size_t len)
-   {
-   m_rng->add_entropy(in, len);
-   }
-
 size_t AutoSeeded_RNG::reseed(Entropy_Sources& srcs,
                               size_t poll_bits,
                               std::chrono::milliseconds poll_timeout)
    {
    return m_rng->reseed(srcs, poll_bits, poll_timeout);
    }
 
-void AutoSeeded_RNG::randomize(uint8_t output[], size_t output_len)
+void AutoSeeded_RNG::fill_bytes_with_input(std::span<uint8_t> out, std::span<const uint8_t> in)
    {
-   m_rng->randomize_with_ts_input(output, output_len);
-   }
+   std::array<uint8_t, 16> additional_input;
+   if(in.empty() && m_rng->accepts_input())
+      {
+      store_le(OS::get_system_timestamp_ns(), additional_input.data());
+      store_le(OS::get_high_resolution_clock(), additional_input.data() + 8);
+      in = additional_input;
```

Comment:
The diff seems to be a bit misleading here. This code merges the behaviours of the former `::randomize()` and `::randomize_with_input()`. Therefore it has a special case for `in` being empty. Nevertheless, we might indeed replace the special case with a call to `randomize_with_ts_input()` and call `randomize_with_input()` in an `else` branch.

This would essentially make the call recursive, though. `randomize_with_ts_input()` will eventually call `fill_bytes_with_input()` again (but not trigger the special case, because `in` won't be empty anymore. I'm frankly not sure whether this little code duplication justifies this.

Sketch:

```C++
void AutoSeeded_RNG::fill_bytes_with_input(std::span<uint8_t> out, std::span<const uint8_t> in)
   {
   if(in.empty())
      {
      // Will call `fill_bytes_with_input()` recursively after filling `in`
      // with entropy pulled from the high resolution system clock.
      m_rng->randomize_with_ts_input(out);
      }
   else
      {
      m_rng->randomize_with_input(out, in);
      }
   }
```

It certainly is tempting, but should we really risk an endless recursion (when `randomize_with_ts_input()` does break its oath)?

Reply:
This is not really recursive though, because `AutoSeeded_RNG` is calling its internal RNG, not itself?

Reply:
Oh! Indeed. What I had in mind when writing that comment before was:

```C++
void AutoSeeded_RNG::fill_bytes_with_input(std::span<uint8_t> out, std::span<const uint8_t> in)
   {
   if(in.empty())
      {
      // Will call `fill_bytes_with_input()` recursively after filling `in`
      // with entropy pulled from the high resolution system clock.
      this->randomize_with_ts_input(out);
   // ^~ !! RECURSE !!
      }
   else
      {
      m_rng->randomize_with_input(out, in);
      }
   }
```

I.e. the call stack would have looked roughly like that:

```
...
application_code()
AutoSeeded_RNG::fill_bytes_with_input(out: <some buffer>, in: {})
AutoSeeded_RNG::fill_bytes_with_ts_input(out: <some buffer>)
AutoSeeded_RNG::fill_bytes_with_input(out: <some buffer>, in: {timestamp}) <= "in" is now filled (recursion abort)
Internal_RNG::fill_bytes_with_input(out: <some buffer>, in: {timestamp})
...
```

But the way I (accidentially) typed out 3 weeks ago would actually make much more sense. Especially with `randomize_with_ts_input()` not being `virtual` anymore. I.e. the internal RNG implementation cannot play any tricks on us. Yikes, object orientation spinning in my head.